include/lbann/models/model.hpp

```diff
@@ -378,23 +353,7 @@ class model {
   // Callbacks
   // ===========================================
 
-  /** @brief Execute callbacks at start of training. */
-  virtual void do_train_begin_cbs();
```

Comment:
I think it's better to keep these functions inside `model` so that the `training_algorithm` doesn't need to touch the callbacks. In this case, we'll need to make the functions public so that the algorithm can call them at the appropriate times.

Reply:
IMO, callbacks, at least those related to training, are not part of the model. For instance, callbacks that trigger at the start of a minibatch require a training algorithm that acknowledges minibatches.

Reply:
I'm wary of splitting these functions across multiple classes. Perhaps it's more natural for callbacks to live in the trainer. The model can activate the per-step callbacks by talking to the trainer at the appropriate times.

Reply:
The callbacks are split into the logical unit that executes them, so these are migrated into the training algorithm.